src/com/puppetlabs/puppetdb/facts.clj

```diff
@@ -171,21 +183,29 @@
                     ;; thus allowing us to store the top level for each
                     ;; fact.
                     fp (if (= depth 0)
-                         (conj mem (factmap-express-node data path))
+                         (conj mem (leaf-fn path data))
                          mem)]
-               (let [new-fp (factmap-to-paths* v fp (conj path k))]
+               (let [new-fp (flatten-facts-with leaf-fn v fp (conj path k))]
                  (if (empty? remaining)
                    new-fp
                    (recur (first remaining)
                           (next remaining)
                           new-fp))))))
          ;; Leaf
-         (conj mem (factmap-express-node data path))))))
+         (conj mem (leaf-fn path data))))))
+
+(defn-validated path->pathmap :- pathmap-scm
+  [path :- fact-path]
+  {:path (factpath-to-string path)
+   :name (first path)
+   :depth (dec (count path))})
 
-(pls/defn-validated factmap-to-paths :- [fact-path-map]
-  "Converts a map of facts to a list of `fact-path-map`s."
-  [hash :- fact-set]
-  (factmap-to-paths* hash))
+;; FIXME: doesn't look like prismatic can handle [[x y]]
```

Comment:
I'm not understanding what you're wanting to handle here with Prismatic, is it just a list of pairs?